python/pyspark/ml/functions.py

```diff
@@ -106,6 +117,474 @@ def array_to_vector(col: Column) -> Column:
+        can be represented by a "sparse" dictionary, where the keys are the integer indices of the
+        inputs, and the values are the shapes.  Each tensor input value in the Spark DataFrame must
+        be represented as a single column containing a flattened 1-D array.  The provided
+        input_tensor_shapes will be used to reshape the flattened array into expected tensor shape.
+        For the list form, the order of the tensor shapes must match the order of the selected
+        DataFrame columns.  The batch dimension (typically -1 or None in the first dimension) should
+        not be included, since it will be determined by the batch_size argument.  Tabular datasets
+        with scalar-valued columns should not provide this argument.
+
+    Returns
+    -------
+    A pandas_udf for predicting a batch.
+    """
+    # generate a new uuid each time this is invoked on the driver to invalidate executor-side cache.
+    model_uuid = uuid.uuid4()
+
+    def predict(data: Iterator[Union[pd.Series, pd.DataFrame]]) -> Iterator[pd.DataFrame]:
+        from pyspark.ml.model_cache import ModelCache
+
+        # get predict function (from cache or from running user-provided predict_batch_fn)
+        predict_fn = ModelCache.get(model_uuid)
+        if not predict_fn:
+            predict_fn = predict_batch_fn()
+            ModelCache.add(model_uuid, predict_fn)
+
+        # get number of expected parameters for predict function
+        signature = inspect.signature(predict_fn)
+        num_expected = len(signature.parameters)
+
+        # convert sparse input_tensor_shapes to dense if needed
+        input_shapes: list[list[int] | None]
+        if isinstance(input_tensor_shapes, Mapping):
+            input_shapes = [None] * num_expected
+            for index, shape in input_tensor_shapes.items():
+                input_shapes[index] = shape
+        else:
+            input_shapes = input_tensor_shapes  # type: ignore
+
+        # iterate over partition, invoking predict_fn with ndarrays
+        for partition in data:
+            has_tuple = isinstance(partition, Tuple)  # type: ignore
+            has_tensors = _has_tensor_cols(partition)
+
+            # require input_tensor_shapes for any tensor columns
+            if has_tensors and not input_shapes:
+                raise ValueError("Tensor columns require input_tensor_shapes")
+
+            for batch in _batched(partition, batch_size):
+                num_actual = len(batch.columns)
+                if num_actual == num_expected and num_expected > 1:
+                    # input column per expected input, convert each column into param
+                    multi_inputs = [batch[col].to_numpy() for col in batch.columns]
+                    if input_shapes:
+                        if len(input_shapes) == num_actual:
+                            multi_inputs = [
+                                np.vstack(v).reshape([-1] + input_shapes[i])  # type: ignore
```

Comment:
Can you add a check that after reshape, the first dimension length equals to `len(batch)` ? If no, raise an error say the input data does not match the provided shape.